File: RansomNote.py
#https://leetcode.com/problems/ransom-note
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        rc=Counter(ransomNote) #Creates a hashmap that keep track of the frequency of each key
        mc=Counter(magazine)

        for ch, c in rc.items(): #ch is the character(key) c is the count(value) of the ransom hashmap
            if mc[ch]<c: # checks if the frequency of the character in the magazine hashmap is lesser than the count of the ransom character
                return False
        return True
        

File: RansomNote.java
//https://leetcode.com/problems/ransom-note
import java.util.HashMap;
import java.util.Map;

public class RansomNote {
    public boolean canConstruct(String ransomNote, String magazine) {
        //Create HashMap to store the frequency of characters from the magazine array
        Map<Character,Integer> map = new HashMap<>();
        for(char c: magazine.toCharArray()){
            map.put(c,map.getOrDefault(c,0)+1); //increment the counter in the map
        }

        //check if the ransomNote can be constructed using the map
        for(char c : ransomNote.toCharArray()){
            //check if character is in the map or not
            if(!map.containsKey(c)||map.get(c)==0){
                return false;
            }
            //decrement character count
            map.put(c,map.get(c)-1);
        }
        return true;
    }
}
